### ISSUE TITLE
Cover all combinations of language and serialization features with tests

### ISSUE THREAD

**Issue by shanshin** [Kotlin/kotlinx.serialization#2324]
A matrix should be drawn up between language features (value classes, abstract classes, sealed classes, nested classes, etc.) and serialization features (generated serializer, custom serializer, contextual serializer, etc.) and simple serialization/deserialization tests should be written for all combinations.

All tests should be implemented in serialization Kotlin plugin, if there is such a possibility.

**Comment by sandwwraith** [comment]
Several notes:

- IIRC `sealed interface` requires `@Serializable` on it to work as sealed (auto-registering children). Otherwise, it works as a regular interface (requires modules). Both of these cases likely should be tested.
- interface inherited class - you mean 'class implementing interface' or what? Interfaces couldn't extend classes last time I've checked)
- For `@Serializable` and `@Contextual` on property: they also can be applied directly to types, e.g. `val foo: List<@Serializable(Custom::class) X>`
- 'Contextual by SerialModule' - not clear what that means. `@Contextual` can be used on property and type, and we also have `@file:UseContextualSerialization`
- All section 'properties' should probably be renamed to 'property types'. They all should be verified, however, the properties itself can come in different flavors: `public/private`, `val/var`, `@Transient`, `@Required`, delegated `by`, etc. It is important to test this flavors.
- 'implicitly contextual property' - do we have those? IIRC there isn't any types for which `@Contextual` applied automatically.
- Descriptors/companions also likely can be accessed from class' own `init {}` block:
```
@Serializable
class X(val f: Foo, val b: Bar) {
    init {
        println(X.serializer().descriptor)
    }

    companion object {
        init {
            println(X.serializer().descriptor)
        }
    }
}
```
- 'Property accessing' - what do you mean by that? Plugin shouldn't affect user-declared properties.

**Comment by shanshin** [comment]
> IIRC sealed interface requires @Serializable on it to work as sealed (auto-registering children). Otherwise, it works as a regular interface (requires modules). Both of these cases likely should be tested.

Yes, this is covered by points `Serializable`: `Generated serializer (@Serializable)` and `Serializable by default (without annotations: enum, interface, sealed interface)`

> interface inherited class - you mean 'class implementing interface' or what? Interfaces couldn't extend classes last time I've checked)

Yes, class implementing interface, I used a common word for uniformity. Maybe an `... , extending ...` will do? E.g. `class, extending interface`

> For @Serializable and @Contextual on property: they also can be applied directly to types, e.g. val foo: List<@Serializable(Custom::class) X>

ok

> 'Contextual by SerialModule' - not clear what that means. @Contextual can be used on property and type, and we also have @file:UseContextualSerialization
> 'implicitly contextual property' - do we have those? IIRC there isn't any types for which @Contextual applied automatically.

indeed, hurried

> Descriptors/companions also likely can be accessed from class' own init {} block:

ok

> 'Property accessing' - what do you mean by that? Plugin shouldn't affect user-declared properties.

ok, I'll remove this


> They all should be verified, however, the properties itself can come in different flavors: public/private, val/var, @Transient, @Required, delegated by, etc. It is important to test this flavors.

- `public/private` I can't imagine how the plugin explicitly affects this
 - `val/var` the finality of the property does not affect the plugin, especially in these test cases it will be difficult to verify the correctness
 - `delegated` I think delegated properties should be checked by another type of tests (which will perform serialization and deserialization)
 - `@Transient, @Required` this only affects the contents of the descriptor. @sandwwraith, should we create tests that completely cover descriptors? (then we need to add annotations, serial names, kinds, optionality, inlining)
